keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,346 @@
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  Spec StorageVersionMigrationSpec
+  Status StorageVersionMigrationStatus
+}
+
+type StorageVersionMigrationSpec {
+  // Resource is the identifier of the resource that's being migrated.
+  // TBD: maybe resource=hash(etcd instance, etcd path prefix).
+  // Immuntable.
+  Resource string 
+  // GroupVersions are all the groups and versions this resource is served by
+  // the apiserver. The server preferred version should be groupVersions[0].
+  // **Migration workers** GET & PUT via the preferred version by default, to
+  // avoid conversions in the apiserver.
+  GroupVersions []GroupVersion
+  // Generation is the generation of the spec. Incrementing generation triggers
+  // migration. We do not use the objectMeta.generation because its semantic is
+  // not clear yet (e.g., objectMeta.generation might be incremented when
+  // labels/annotations are updated [#67428][])
+  Generation int64
+  // Reason of the migration. Examples are "manually triggered", "apiserver
+  // updated", etc.
+  Reason string
+}
+
+type GroupVersion struct {
+  Group string
+  Version string
+}
+
+type StorageVersionMigrationStatus {
+  // ObeservedGeneration is the spec generation that has been observed by the
+  // **migration controller** It indicates a **migration worker** for this
+  // generation has been launched.
+  ObeservedGeneration int64
+  // CompletedGeneration is the greatest spec generation whose **migration worker**
+  // has completed.
+  CompletedGeneration int64
+}
+```
+
+[#67428]:https://github.com/kubernetes/kubernetes/issues/67428
+[continue token]:https://github.com/kubernetes/kubernetes/blob/972e1549776955456d9808b619d136ee95ebb388/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L82
+[workflow section]:#alpha-workflow
+
+We need to expand the discovery API to recognize APIs in different groups
+that share the same storage, e.g., `extensions/v1beta1/deployment` and
+`apps/v1/deployment` share the same key space in etcd. Specifically, we expand
+the [APIResource][] API:
+
+```golang
+type APIResource struct {
+  Name string
+  ...
+  // StorageUID=SHA-256(etcd server in the form of scheme://ip:port, etcd path),
+  // obfuscated with a random key. E.g., for pods, it is
+  // SHA-256(scheme://ip:port/registry/pods).
+  // The string is obfuscated to avoid exposing server configuration details.
+  StorageUID string
+}
+```
+
+[APIResource]:https://github.com/kubernetes/kubernetes/blob/d38fb2c6451447ad0ecbdfed971a563419d60883/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L874
+
+
+### Failure recovery
+
+As stated in the goals section, the migration has to make progress even if the
+environment is flaky. This section describes how the migrator recovers from
+failure.
+
+Kubernetes **replicaset controller** restarts the **migration controller** `pod`
+if it fails. When the **migration controller** starts, it resumes migration
+by reading the existing `migration` objects.
```

Comment:
i'm still confused. Below, it lists all the tasks done at the "kube-migrator controller" start-up time. My question is how does it know it is the first time (started by kube-migrator command) vs a second time (it is killed and restarted by the replicaset controller)

==================================================
On start-up, the kube-migrator controller does the following,

discovers all supported APIs.
lists all custom resource definitions, excludes them from the API list retrieved in the previous step.
for each non-custom resource, creates a migration CR (see the API section for the schema).
for each migration instance, increments the migration.spec.generation, and logs the reason.

Reply:
Sorry for the confusion.

Updated this section to:
```
Kubernetes **replicaset controller** restarts the **migration controller** `pod`
if it fails. The restarted **migration controller** follows the same bootstrap
workflow as described in the [workflow section][]. Because the **migration
controller** adopts existing `migration` objects, it can resume from where
it left off.
```

and updated the workflow section to
```
On start-up, the **kube-migrator controller** does the following,
...
* for each non-custom resource, creates a `migration` CR (see the [API
  section][] for the schema) if it doesn't exist already.
* for each newly created `migration` instance, increments the
  `migration.spec.generation`, and logs the reason.
...
```